### PR TITLE
Fix undo manager

### PR DESCRIPTION
**Pull request by amolenaar**
### PR Type
What kind of change does this PR introduce?

- [X] Bugfix
- [X] Feature
- ~Code style update (formatting, local variables)~
- [X] Refactoring (no functional changes, no api changes)
-  ~Documentation content changes~

### What is the current behavior?

Undo behavior is a bit broken when using Gaphas 3.

I ran into some trouble that made be decide to recreate elements, instead of reusing them, not sure what that was though. Recreating elements does render Gaphas' state recorder useless.

If we drop the state recorder, new events have to be created for:
- handle (position) changes
- matrix updates
- connect/disconnect
- line splitting and merging

Issue Number: #585 

### What is the new behavior?

- [X] Undo manager stores undo data using element id's, not the element itself.
- [X]  Fix ObjectNodeItem's show ordering property
- [X] Elements and diagram items are recreated on undo/redo, instead of reused.
- [X] No longer use Gaphas' `state` module
  - [X] Fix update behavior for matrix and Handle updates -> special "reversible events"
  - [X] Special events for connect and disconnect
  - [X] Special events for split and merge of line segments
  - [x] Only handle/position updates are allowed to happen outside of a transaction (due to constraint solving)
  - [x] All handles are watched for state changes
- [X] ElementDispatcher no longer catches exceptions thrown from handlers
- [X] attribute and association properties revert their state when an exception is raised from a handler
- [X] Diagram can not be updated on a Presentation (because constraints are connected to the diagram's Connections instance)
- [X] Simple items (line, box, ellipse) inherit from LinePresentation and ElementPresentation now -> reduce code
- [x] Issue: Undo manager does not play well with the UML Sanitizer service -> bypass sanitizer when undoing a transaction?
- [X] Diagram item state is not properly restores (after paste item ends up in top-left corner) -> should save values  of old item. When an item is removed, no updates for matrix, handles, etc. are recorded. Hence, they should be stored on the undo event.
- [x] Remove concept of "transient" model elements. This does not play well with undo.

### CODE REVIEW

File: gaphor/UML/sanitizerservice.py [hunk 1]
```diff
@@ -10,12 +10,21 @@
...
+    return guard
+
+
 class SanitizerService(Service):
     """Does some background cleanup jobs, such as removing elements from the
     model that have no presentations (and should have some)."""
 
-    def __init__(self, event_manager):
+    def __init__(self, event_manager, undo_manager=None):
         self.event_manager = event_manager
```
Comment (amolenaar) [inline gaphor/UML/sanitizerservice.py:26]: I made the undo-manager optional, to avoid braking all TestCase unit tests.

**Review by danyeaw — COMMENTED** [review]
This looks really good, I think this is going to greatly improve the undo system's robustness! I have a few minor comments and I found an issue while testing it that I will also post.

**Comment by danyeaw** [comment]
I am getting a series of AttributeErrors and then a ValueError if I try to create a composite or shared association.

To reproduce:

1. Create two classes
2. Try to create a composite association between them

```
gaphor.transaction ERROR Transaction terminated due to an exception, performing a rollback
Traceback (most recent call last):
  File "/home/dan/Projects/gaphor/gaphor/transaction.py", line 29, in _transactional
    return func(*args, **kwargs)
  File "/home/dan/Projects/gaphor/gaphor/UML/classes/classespropertypages.py", line 556, in _on_end_navigability_change
    UML.model.set_navigability(
  File "/home/dan/Projects/gaphor/gaphor/UML/modelfactory.py", line 339, in set_navigability
    owner.ownedAttribute = end
AttributeError: 'NoneType' object has no attribute 'ownedAttribute'
gaphor.diagram.diagramtools.connector ERROR Error during connect
Traceback (most recent call last):
  File "/home/dan/Projects/gaphor/gaphor/diagram/diagramtools/connector.py", line 65, in connect
    adapter.connect(handle, sink.port)
  File "/home/dan/Projects/gaphor/gaphor/diagram/connectors.py", line 308, in connect
    self.connect_subject(handle)
  File "/home/dan/Projects/gaphor/gaphor/UML/classes/classconnect.py", line 137, in connect_subject
    line.head_subject.type = c1.subject  # type: ignore[assignment]
  File "/home/dan/Projects/gaphor/gaphor/core/modeling/properties.py", line 142, in __set__
    self._set(obj, value)
  File "/home/dan/Projects/gaphor/gaphor/core/modeling/properties.py", line 381, in _set
    self._set_one(obj, value, from_opposite)
  File "/home/dan/Projects/gaphor/gaphor/core/modeling/properties.py", line 409, in _set_one
    self.handle(AssociationSet(obj, self, old, value))
  File "/home/dan/Projects/gaphor/gaphor/core/modeling/properties.py", line 174, in handle
    event.element.handle(event)
  File "/home/dan/Projects/gaphor/gaphor/core/modeling/element.py", line 141, in handle
    model.handle(event)
  File "/home/dan/Projects/gaphor/gaphor/core/modeling/elementfactory.py", line 188, in handle
    self.event_manager.handle(event)
  File "/home/dan/Projects/gaphor/gaphor/core/eventmanager.py", line 55, in handle
    self._events.handle(e)
  File "/home/dan/Projects/gaphor/.venv/lib/python3.9/site-packages/generic/event.py", line 64, in handle
    handler(event)
  File "/home/dan/Projects/gaphor/gaphor/core/modeling/elementdispatcher.py", line 241, in on_element_change_event
    handler(event)
  File "/home/dan/Projects/gaphor/gaphor/UML/classes/classespropertypages.py", line 509, in restore_nav_handler
    self._on_end_navigability_change(combo, end)
  File "/home/dan/Projects/gaphor/gaphor/transaction.py", line 29, in _transactional
    return func(*args, **kwargs)
  File "/home/dan/Projects/gaphor/gaphor/UML/classes/classespropertypages.py", line 556, in _on_end_navigability_change
    UML.model.set_navigability(
  File "/home/dan/Projects/gaphor/gaphor/UML/modelfactory.py", line 339, in set_navigability
    owner.ownedAttribute = end
AttributeError: 'NoneType' object has no attribute 'ownedAttribute'
gaphor.transaction ERROR Transaction terminated due to an exception, performing a rollback
Traceback (most recent call last):
  File "/home/dan/Projects/gaphor/gaphor/transaction.py", line 29, in _transactional
    return func(*args, **kwargs)
  File "/home/dan/Projects/gaphor/gaphor/UML/classes/classespropertypages.py", line 557, in _on_end_navigability_change
    end.subject.association, end.subject, self.NAVIGABILITY[combo.get_active()]
AttributeError: 'NoneType' object has no attribute 'association'
gaphor.services.undomanager ERROR Error while undoing action <gaphor.UML.uml.Association element 2221fa72-52f1-11eb-b403-18cf5efc6bb9>.memberEnd delete <gaphor.UML.uml.Property element 22239c9c-52f1-11eb-b403-18cf5efc6bb9>.
Traceback (most recent call last):
  File "/home/dan/Projects/gaphor/gaphor/services/undomanager.py", line 73, in execute
    act()
  File "/home/dan/Projects/gaphor/gaphor/services/undomanager.py", line 445, in _undo_association_add_event
    association._del(element, value, from_opposite=True)
  File "/home/dan/Projects/gaphor/gaphor/core/modeling/properties.py", line 449, in _del
    self._del_many(obj, value, from_opposite, do_notify)
  File "/home/dan/Projects/gaphor/gaphor/core/modeling/properties.py", line 481, in _del_many
    self.handle(AssociationDeleted(obj, self, value))
  File "/home/dan/Projects/gaphor/gaphor/core/modeling/properties.py", line 174, in handle
    event.element.handle(event)
  File "/home/dan/Projects/gaphor/gaphor/core/modeling/element.py", line 141, in handle
    model.handle(event)
  File "/home/dan/Projects/gaphor/gaphor/core/modeling/elementfactory.py", line 188, in handle
    self.event_manager.handle(event)
  File "/home/dan/Projects/gaphor/gaphor/core/eventmanager.py", line 55, in handle
    self._events.handle(e)
  File "/home/dan/Projects/gaphor/.venv/lib/python3.9/site-packages/generic/event.py", line 64, in handle
    handler(event)
  File "/home/dan/Projects/gaphor/gaphor/core/modeling/elementdispatcher.py", line 241, in on_element_change_event
    handler(event)
  File "/home/dan/Projects/gaphor/gaphor/UML/classes/classespropertypages.py", line 509, in restore_nav_handler
    self._on_end_navigability_change(combo, end)
  File "/home/dan/Projects/gaphor/gaphor/transaction.py", line 29, in _transactional
    return func(*args, **kwargs)
  File "/home/dan/Projects/gaphor/gaphor/UML/classes/classespropertypages.py", line 557, in _on_end_navigability_change
    end.subject.association, end.subject, self.NAVIGABILITY[combo.get_active()]
AttributeError: 'NoneType' object has no attribute 'association'
gaphor.transaction ERROR Transaction terminated due to an exception, performing a rollback
Traceback (most recent call last):
  File "/home/dan/Projects/gaphor/gaphor/transaction.py", line 29, in _transactional
    return func(*args, **kwargs)
  File "/home/dan/Projects/gaphor/gaphor/UML/classes/classespropertypages.py", line 557, in _on_end_navigability_change
    end.subject.association, end.subject, self.NAVIGABILITY[combo.get_active()]
AttributeError: 'NoneType' object has no attribute 'association'
gaphor.services.undomanager ERROR Error while undoing action <gaphor.UML.uml.Association element 2221fa72-52f1-11eb-b403-18cf5efc6bb9>.memberEnd delete <gaphor.UML.uml.Property element 222326c2-52f1-11eb-b403-18cf5efc6bb9>.
Traceback (most recent call last):
  File "/home/dan/Projects/gaphor/gaphor/services/undomanager.py", line 73, in execute
    act()
  File "/home/dan/Projects/gaphor/gaphor/services/undomanager.py", line 445, in _undo_association_add_event
    association._del(element, value, from_opposite=True)
  File "/home/dan/Projects/gaphor/gaphor/core/modeling/properties.py", line 449, in _del
    self._del_many(obj, value, from_opposite, do_notify)
  File "/home/dan/Projects/gaphor/gaphor/core/modeling/properties.py", line 481, in _del_many
    self.handle(AssociationDeleted(obj, self, value))
  File "/home/dan/Projects/gaphor/gaphor/core/modeling/properties.py", line 174, in handle
    event.element.handle(event)
  File "/home/dan/Projects/gaphor/gaphor/core/modeling/element.py", line 141, in handle
    model.handle(event)
  File "/home/dan/Projects/gaphor/gaphor/core/modeling/elementfactory.py", line 188, in handle
    self.event_manager.handle(event)
  File "/home/dan/Projects/gaphor/gaphor/core/eventmanager.py", line 55, in handle
    self._events.handle(e)
  File "/home/dan/Projects/gaphor/.venv/lib/python3.9/site-packages/generic/event.py", line 64, in handle
    handler(event)
  File "/home/dan/Projects/gaphor/gaphor/core/modeling/elementdispatcher.py", line 241, in on_element_change_event
    handler(event)
  File "/home/dan/Projects/gaphor/gaphor/UML/classes/classespropertypages.py", line 509, in restore_nav_handler
    self._on_end_navigability_change(combo, end)
  File "/home/dan/Projects/gaphor/gaphor/transaction.py", line 29, in _transactional
    return func(*args, **kwargs)
  File "/home/dan/Projects/gaphor/gaphor/UML/classes/classespropertypages.py", line 557, in _on_end_navigability_change
    end.subject.association, end.subject, self.NAVIGABILITY[combo.get_active()]
AttributeError: 'NoneType' object has no attribute 'association'
gaphor.services.undomanager ERROR Error while undoing action Undo create element <gaphor.UML.uml.Association element 2221fa72-52f1-11eb-b403-18cf5efc6bb9>.
Traceback (most recent call last):
  File "/home/dan/Projects/gaphor/gaphor/services/undomanager.py", line 73, in execute
    act()
  File "/home/dan/Projects/gaphor/gaphor/services/undomanager.py", line 338, in _undo_create_event
    element = self.deep_lookup(element_id)
  File "/home/dan/Projects/gaphor/gaphor/services/undomanager.py", line 282, in deep_lookup
    raise ValueError(f"Element with id {id} not found in model")
ValueError: Element with id 2221fa72-52f1-11eb-b403-18cf5efc6bb9 not found in model
gaphor.services.undomanager ERROR Error while undoing action <gaphor.UML.uml.Association element 2221fa72-52f1-11eb-b403-18cf5efc6bb9>.presentation add <gaphor.UML.classes.association.AssociationItem element 22226d9a-52f1-11eb-b403-18cf5efc6bb9>.
Traceback (most recent call last):
  File "/home/dan/Projects/gaphor/gaphor/services/undomanager.py", line 73, in execute
    act()
  File "/home/dan/Projects/gaphor/gaphor/services/undomanager.py", line 463, in _undo_association_delete_event
    element = self.deep_lookup(element_id)
  File "/home/dan/Projects/gaphor/gaphor/services/undomanager.py", line 282, in deep_lookup
    raise ValueError(f"Element with id {id} not found in model")
ValueError: Element with id 2221fa72-52f1-11eb-b403-18cf5efc6bb9 not found in model
```

File: gaphor/services/tests/test_undo_presentation.py [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+
+def test_line_create(diagram, undo_manager, event_manager, capsys):
+    with Transaction(event_manager):
+        LinePresentation(diagram)
+
+    assert diagram.ownedPresentation
+
+    undo_manager.undo_transaction()
+    # Errors in undo manager are reported on stdout
+    captured = capsys.readouterr()
```
[the remark this answers is not in the document]
Comment (amolenaar) [reply gaphor/services/tests/test_undo_presentation.py:27]: Actually, there's also a `caplog` fixture, which is even nicer: it hooks into the `logging` module. I should probably change this to use caplog instead.

**Comment by amolenaar** [comment]
@danyeaw Thanks for the review :+1: I'll look into the issue with composite associations.

File: gaphor/core/modeling/element.py [hunk 3]
```diff
@@ -65,7 +57,7 @@ def __init__(
         A model can be provided to refer to the model this element belongs to.
         """
         super().__init__()
-        self._id: Id = id or (id is not False and str(uuid.uuid1()) or False)
+        self._id: Id = id or str(uuid.uuid1())
```
[the remark this answers is not in the document]
Comment (amolenaar) [reply gaphor/core/modeling/element.py:60]: Nice catch!

**Comment by amolenaar** [comment]
I fixed the error. Introduced it as well :see_no_evil:

**Comment by danyeaw** [comment]
It looks like everything is working great now. Really well done, undo is working great now! :tada:

**Pull request merged**